### PR TITLE
Change linktype raw to 101 to match http://www.tcpdump.org/linktypes.…

### PR DESCRIPTION
**Pull request by haufes**
Change linktype raw to 101 to match http://www.tcpdump.org/linktypes.html
I left other link types in, but I'm not sure how valid they are.

As well added a packet stream reader to the test project that can deal correctly with endiannes in the pcap files.

### CODE REVIEW

**Comment by haufes** [comment]
This will resolve this issue https://github.com/chmorgan/packetnet/issues/57

**Comment by Phyxion** [comment]
Can you create a PR with just the link layer changes? The other changes can be in a separate PR to make or easier to review. I also think the changes in Packet.cs and PacketDotNet.csproj are not required. I'll look over the test stuff in the other PR then.

Thanks!
On Dec 17, 2018, at 19:12, Steven Haufe <notifications@github.com<mailto:notifications@github.com>> wrote:

Change linktype raw to 101 to match http://www.tcpdump.org/linktypes.html
I left other link types in, but I'm not sure how valid they are.

As well added a packet stream reader to the test project that can deal correctly with endiannes in the pcap files.

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/chmorgan/packetnet/pull/82

Commit Summary

  *   Change linktype raw to 101 to match http://www.tcpdump.org/linktypes.html

File Changes

  *   M PacketDotNet/LinkLayers.cs<https://github.com/chmorgan/packetnet/pull/82/files#diff-0> (13)
  *   M PacketDotNet/Packet.cs<https://github.com/chmorgan/packetnet/pull/82/files#diff-1> (4)
  *   M PacketDotNet/PacketDotNet.csproj<https://github.com/chmorgan/packetnet/pull/82/files#diff-2> (1)
  *   A Test/ByteSwappedBinaryReader.cs<https://github.com/chmorgan/packetnet/pull/82/files#diff-3> (134)
  *   M Test/PacketType/ICMPv4PacketTest.cs<https://github.com/chmorgan/packetnet/pull/82/files#diff-4> (15)
  *   M Test/PacketType/IPv6PacketTest.cs<https://github.com/chmorgan/packetnet/pull/82/files#diff-5> (47)
  *   A Test/PcapHeader.cs<https://github.com/chmorgan/packetnet/pull/82/files#diff-6> (33)
  *   A Test/PcapMagic.cs<https://github.com/chmorgan/packetnet/pull/82/files#diff-7> (35)
  *   A Test/PcapPacket.cs<https://github.com/chmorgan/packetnet/pull/82/files#diff-8> (71)
  *   A Test/PcapStream.cs<https://github.com/chmorgan/packetnet/pull/82/files#diff-9> (28)
  *   M Test/Test.csproj<https://github.com/chmorgan/packetnet/pull/82/files#diff-10> (5)

Patch Links:

  *   https://github.com/chmorgan/packetnet/pull/82.patch
  *   https://github.com/chmorgan/packetnet/pull/82.diff

**Comment by Phyxion** [comment]
Can you create a PR with just the link layer changes? The other changes can be in a separate PR to make or easier to review. I also think the changes in Packet.cs and PacketDotNet.csproj are not required. I'll look over the test stuff in the other PR then. Thanks!

**Comment by haufes** [comment]
I've pulled out the update to the Packet.cs and PacketDotNet.csproj, they where just to allow me to debug in VS, but I've left in the other files. It looks like Pcaplib isn't dealing with the endiannes  correctly in the raw.pcap files. I needed to add a packet stream reader that would work, other wise the tests would fail.

File: Test/PcapStream.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+﻿using System.IO;
+namespace Test
+{
+    public class PcapStream
+    {
+        BinaryReader buffer;
```
Comment (Phyxion) [inline Test/PcapStream.cs]: Can you fix the naming to be similar to the style in the rest of the project?
So private fields should be named with private and _likeThis.
public fields should be Capitalized. This also applies to the other files.

**Comment by Phyxion** [comment]
Left a few comments, rest looks good.

**Comment by haufes** [comment]
Hope I got them all

**Comment by Phyxion** [comment]
Left a few more comments that should be easy to address, the rest looks good to me. If @chmorgan can give his OK I'll merge it then.

**Comment by haufes** [comment]
@Phyxion  I don't see any new comments.

**Comment by Phyxion** [comment]
@haufes They are now above in the review, I forgot to hit the submit button 😢

**Comment by haufes** [comment]
@Phyxion I've refactored the code a bit to address your comment

File: Test/PacketType/IPv6PacketTest.cs [hunk 2]
```diff
@@ -70,28 +66,33 @@ public void VerifyPacket0(Packet p, RawCapture rawCapture, LinkLayers linkLayer)
...
+                while (p != null)
                 {
-                case 0:
-                    VerifyPacket0(p, rawCapture, linkLayer);
-                    break;
-                default:
-                    Assert.Fail("didn't expect to get to packetIndex " + packetIndex);
-                    break;
-                }
+                    Console.WriteLine("got packet");
```
Comment (Phyxion) [inline Test/PacketType/IPv6PacketTest.cs]: Should be removed (added in last commit) 👍

File: Test/PcapHeader.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        public ushort MinorVersion { get; } = 0;
+        public uint ThisZone { get; } = 0;
+        public uint SigFigs { get; } = 0;
+        public uint SnapLen { get; } = 0;
+        public LinkLayers LinkType { get; } = 0;
+
+        public PcapHeader(BinaryReader readBuffer)
+        {
+            byte[] Buffer;
+            Buffer = new byte[20];
```
Comment (Phyxion) [inline Test/PcapHeader.cs]: => var buffer = new byte[20]

File: Test/PcapHeader.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+            readBuffer.Read(Buffer, 0, 20);
+
+            this.MajorVersion = BitConverter.ToUInt16(Buffer, 0);
```
Comment (Phyxion) [inline Test/PcapHeader.cs]: => remove this

File: Test/PcapMagic.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+using System.IO;
+
+namespace Test
+{
+    public class PcapMagic
+    {
+        const uint PCAP_MAGIC_NUMBER = 0xa1b2c3d4;
+        const uint ALTERNATE_PCAP_MAGIC_NUMBER = 0xa1b2cd34; //this one shows up in some older captures
+
+        public bool ByteInOrder {get; } = true;
```
Comment (Phyxion) [inline Test/PcapMagic.cs]: Add space

File: Test/PcapPacket.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using PacketDotNet;
+using System.IO;
+
+namespace Test
+{
+    public class PcapPacket
+    {
+        static int PACKET_HEADER_LENGTH = 16;
+        public Packet Packet;
```
Comment (Phyxion) [inline Test/PcapPacket.cs]: Add { get; }

**Comment by Phyxion** [comment]
Thanks, added a few nits, but looks good for the rest 👍

**Comment by haufes** [comment]
@Phyxion did you submit them? :)

**Comment by Phyxion** [comment]
Sigh, yes, it's now submitted 👍

File: Test/PcapPacket.cs (hunk 6 above)
Comment (haufes) [reply Test/PcapPacket.cs]: Can't, it's being set from the static method

Comment (Phyxion) [reply Test/PcapPacket.cs]: Ah, yes, you can do {get; private set;}

**Comment by Phyxion** [comment]
@haufes Looking at pcap files now in lib pcap, I think this can and should be addressed within sharppcap (which should be a rather trivial change as there's a pcap_is_swapped function). That means we can continue to use the existing code to read, rather than adding another reader to this project. In the mean time, until this is fixed, the few additional pcap files that need to be added can be manually converted to correct endianness. I've committed a change to the link types.

**Comment by Phyxion** [comment]
(just cleaning up a little on the comments, loading this page took about a minute here 👎 )

**Pull request closed without merging**